Dp24 ancestral plots

DLBPointon opened the pull request:
Tested locally

Adds Ancestral Element Analysis - new requirement for GenomeNote from Psyche, it is also expected to expand to more clades too.

This isn't final as genomnote.nf - line 181 needs a look - Is this the right way to go about it?

BethYates committed:
Merge pull request #146 from sanger-tol/dev

Release 2.0.0

tkchafin committed:
Merge pull request #161 from sanger-tol/dev

2.1.0 release

muffato committed:
Merge pull request #175 from sanger-tol/dev

Release 2.1.1

muffato committed:
Merge branch 'dev'

DLBPointon committed:
initial additions for ancestral subworkflow

DLBPointon committed:
initial additions for ancestral subworkflow

DLBPointon committed:
Adding Annotation_Ancestral into the pipeline

DLBPointon committed:
update

DLBPointon committed:
Adding the scripts

DLBPointon committed:
Linting

DLBPointon committed:
Merge branch 'dev' into dp24_ancestral_plots

DLBPointon committed:
Linting updates

muffato commented on subworkflows/local/genome_statistics.nf:
```diff
@@ -232,6 +232,7 @@ workflow GENOME_STATISTICS {
     ch_summary          = GENESCOPEFK.out.summary                   // channel: [ meta, summary ]
     ch_trans_lin_plot   = GENESCOPEFK.out.transformed_linear_plot   // channel: [ meta, transformed_linear_plot ]
     ch_trans_log_plot   = GENESCOPEFK.out.transformed_log_plot      // channel: [ meta, transformed_log_plot ]
+    busco_full_table = BUSCO.out.full_table          // channel: [ meta, busco_dir ]
```
Since you carefully aligned everything else ;)
```suggestion
    busco_full_table    = BUSCO.out.full_table                      // channel: [ meta, busco_dir ]
```

DLBPointon replied:
😱 Fixed, along with the other alignments I missed.

muffato commented on nextflow.config:
I'm curious. Why is everything shifted right, here ? It looked already aligned 🤔

DLBPointon replied:
Honestly, no idea, case of I started tabbing and didn't stop. I've cleaned it up now.

muffato commented on CHANGELOG.md:
```diff
@@ -3,32 +3,39 @@
+| Old parameter | New parameter     |
+| ------------- | ----------------- |
+|               | --busco_lineage   |
+|               | --ancestral_tabls |
```
```suggestion
|               | --ancestral_table |
```

muffato commented on CHANGELOG.md:
```diff
@@ -3,32 +3,39 @@
+| `cat`                 |             | 2.3.4            |
+| `genescopefk`         |             | 1.2              |
+| `fastk_histex`        |             | 1.1.0            |
+| `samtools/faidx`      |             | 1.21--h50ea8bc_0 |
```
We don't usually report the build number
```suggestion
| `samtools/faidx`      |             | 1.21 |
```

DLBPointon replied:
Done

muffato commented on modules/local/ancestral/plot/environment.yml:
```diff
@@ -0,0 +1,9 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/environment-schema.json
+channels:
+  - conda-forge
+  - bioconda
+dependencies:
+  - conda-forge::r-base=4.1.3
+  - conda-forge::r-tidyverse=1.3.1
+  - bioconda::r-optparse=1.7.3
```
`r-optparse` is in `conda-forge`, not `bioconda`

DLBPointon replied:
Done

muffato commented on modules/local/ancestral/plot/main.nf:
```diff
@@ -0,0 +1,51 @@
+    conda "${moduleDir}/environment.yml"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'oras://community.wave.seqera.io/library/r-optparse_r-scales_r-tidyverse:3285e4530d808c27' :
+        'community.wave.seqera.io/library/r-optparse_r-scales_r-tidyverse:9b4155b6beecffd1' }"
```
The name doesn't match the Conda file (`r-scales` vs `r-optparse`)

DLBPointon replied:
Added in now r-scales=1.4.0

muffato commented on modules/local/ancestral/plot/main.nf:
```diff
@@ -0,0 +1,51 @@
+
+    conda "${moduleDir}/environment.yml"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'oras://community.wave.seqera.io/library/r-optparse_r-scales_r-tidyverse:3285e4530d808c27' :
```
We don't want to use `oras://` containers because they're not supported by older versions of Singularity, cf https://nf-co.re/blog/2024/seqera-containers-part-2#singularity-oras-or-https

DLBPointon replied:
No problem, i ended up using oras because i couldn't get the https one working. I'll test it again.

DLBPointon replied:
Testing is fine this time ¯\_(ツ)_/¯

muffato commented on subworkflows/local/annotation_ancestral.nf:
```diff
@@ -0,0 +1,64 @@
+
+workflow ANNOTATION_ANCESTRAL {
+    take:
+    reference_tuple      // Channel: [ meta, datafile ]
```
Rename to something like `fasta` and change the comment so that it's clear we're talking about the genome

DLBPointon replied:
Done

DLBPointon commented:
I've addressed everything above + I've updated the pipeline to execute the annotation_ancestral on whether an ancestral table has been provided. Decoupling it from lineage.

Removed the table from assets and instead added a link to the raw tsv on github. The official publication docs are versioned here:

https://github.com/charlottewright/Chromosome_evolution_Lepidoptera_MS/blob/pub_v1/sup_tables/TableS4_Merian_element_definitions.tsv

Testing is taking a longer than usual as I'm getting ncbi `too many request` errors - They suggest we start using API keys.

DLBPointon committed:
Update for comments from both inperson and github coversations

DLBPointon committed:
Remove table

DLBPointon committed:
Linting

DLBPointon committed:
Lintint

DLBPointon committed:
Re-adding ancestral_table

DLBPointon commented:
#182

Tested locally with an api-key from ncbi, pipeline runs through.
Tested with ancestral_table - Works
Tested without ancestral_table - Works with the caveat that SAMTOOLS_FAIDX still runs, though I think this is ok for this PR as the surrounding PRs will require that samtools is brought into the genomenote.nf workflow.

Use a GitHub secret to pass it through to the pipeline as a param? Only issue will be If the pipeline fails on that step anyway it will print out the value.

DLBPointon committed:
Update to docs

DLBPointon committed:
Merge branch 'dev' into dp24_ancestral_plots

DLBPointon committed:
Prettier linting

DLBPointon committed:
Merge branch 'dev' into dp24_ancestral_plots

muffato commented:
HI @DLBPointon . I'm having an issue when running the pipeline. It can't find the Singularity image:
```
ERROR ~ Error executing process > 'SANGERTOL_GENOMENOTE:GENOMENOTE:ANNOTATION_ANCESTRAL:ANCESTRAL_PLOT (1)'

Caused by:
  Failed to pull singularity image
    command: singularity pull  --name community.wave.seqera.io-library-r-optparse_r-scales_r-tidyverse-3285e4530d808c27.img.pulling.1747612192822 https://community.wave.seqera.io/library/r-optparse_r-scales_r-tidyverse:3285e4530d808c27 > /dev/null
    status : 255
    hint   : Try and increase singularity.pullTimeout in the config (current is "20m")
    message:
      INFO:    Downloading network image
      FATAL:   the requested image was not found
```

Indeed, the URL leads nowhere
```
mm49@farm22-head1:~ $ wget https://community.wave.seqera.io/library/r-optparse_r-scales_r-tidyverse:3285e4530d808c27
--2025-05-19 09:50:16--  https://community.wave.seqera.io/library/r-optparse_r-scales_r-tidyverse:3285e4530d808c27
Resolving community.wave.seqera.io (community.wave.seqera.io)... 13.41.29.181, 13.134.64.167
Connecting to community.wave.seqera.io (community.wave.seqera.io)|13.41.29.181|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /repo/library/r-optparse_r-scales_r-tidyverse:3285e4530d808c27 [following]
--2025-05-19 09:50:17--  https://community.wave.seqera.io/repo/library/r-optparse_r-scales_r-tidyverse:3285e4530d808c27
Reusing existing connection to community.wave.seqera.io:443.
HTTP request sent, awaiting response... 404 Not Found
2025-05-19 09:50:17 ERROR 404: Not Found.
```

DLBPointon commented:
> HI @DLBPointon . I'm having an issue when running the pipeline. It can't find the Singularity image:
> 
> 
> Indeed, the URL leads nowhere
> 

Hi @muffato,
That's disappointing, I had it working last week.

What are these pointing to then? https://community.wave.seqera.io/repo/library/r-optparse_r-scales_r-tidyverse

Building a singularity image here: https://seqera.io/containers/?packages=R::r-optparse=1.7.3+R::r-scales=1.2.1+R::r-tidyverse=2.0.0


Edit: Looks like there is a dependency that fails to be resolved.

muffato commented:
So far the Singularity URLs I've seen are very different from the Docker ones, cf https://github.com/sanger-tol/genomenote/blob/main/modules/nf-core/samtools/view/main.nf#L7-L8

`community.wave.seqera.io/library/r-optparse_r-scales_r-tidyverse:9b4155b6beecffd1` is valid for Docker containers, and for Singularity as long as it's just that (because Singularity is able to pull from Docker registries) but plugging `https://` won't work because the way docker containers are access is not http.

I don't know if there's a way of recovering the Singularity URL of an already-built container.

DLBPointon committed:
update module

DLBPointon committed:
Merge branch 'dp24_ancestral_plots' of https://github.com/sanger-tol/genomenote into dp24_ancestral_plots

DLBPointon committed:
Updating requirements for plot

DLBPointon committed:
Merge branch 'dp24_ancestral_plots' of https://github.com/sanger-tol/genomenote into dp24_ancestral_plots

DLBPointon commented:
@muffato I've built a sanger-tol container and looks like both docker and singularity are working well and generating output.

One issue I've noticed on the runner though is that the ENA requests fail when both test pipelines are running at the same time. Hopefully this is temporary or may more of the API calls need the errorStrategy adding.

muffato commented:
> One issue I've noticed on the runner though is that the ENA requests fail when both test pipelines are running at the same time. Hopefully this is temporary or may more of the API calls need the errorStrategy adding.

Yeah, I've seen some of those jobs hanging this week. Maybe because of the data centre outage, but still, it made me think.
Since we're not going to use this part for internal genome notes (there's GNKore instead), what about adding a flag to control part of that subworkflow ? Maybe tie it to `note_template` being set.

DLBPointon commented:
> Yeah, I've seen some of those jobs hanging this week. Maybe because of the data centre outage, but still, it made me think. Since we're not going to use this part for internal genome notes (there's GNKore instead), what about adding a flag to control part of that subworkflow ? Maybe tie it to `note_template` being set.

Thats fair yeah, in another PR?

So do something on the lines of:
```
If (note_template) { GENOME_METADATA(stuff...)}
```

and remove `note_template` from test.config?

muffato commented:
Yes, in another PR.

`if (note_template) {`: possibly, but I can see in the workflow file that `GENOME_METADATA.out` is used for the `COMBINE_NOTE_DATA` subworkflow, which is essentially using `ch_note_template` to control running.
Whichever way works :)

DLBPointon committed:
Update module

DLBPointon committed:
minor update for alignment

muffato commented:
The R script doesn't want to run on Singularity:
```
ERROR ~ Error executing process > 'SANGERTOL_GENOMENOTE:GENOMENOTE:ANNOTATION_ANCESTRAL:ANCESTRAL_PLOT (GCA_963859965.1)'

Caused by:
  Process `SANGERTOL_GENOMENOTE:GENOMENOTE:ANNOTATION_ANCESTRAL:ANCESTRAL_PLOT (GCA_963859965.1)` terminated with an error exit status (127)


Command executed:

  plot_buscopainter.R \
      -f buscopainter_complete_location.tsv \
      -p GCA_963859965.1 \
      -i GCA_963859965.1.fasta.fai \
      -m True -d True
  
  cat <<-END_VERSIONS > versions.yml
  "SANGERTOL_GENOMENOTE:GENOMENOTE:ANNOTATION_ANCESTRAL:ANCESTRAL_PLOT":
      R: $(R --version | grep -oP "\d+\.\d+\.\d+")
      plot_buscopainter.R: $(plot_buscopainter.R -v)
  END_VERSIONS

Command exit status:
  127

Command output:
  (empty)

Command error:
  /usr/bin/env: ‘Rscript’: No such file or directory

Work dir:
  /lustre/scratch124/tol/teams/tolit/users/mm49/nextflow/genomenote/work/f6/665f5b2c1785a8f9e2d6b27bb78fc1

Container:
  /nfs/users/nfs_m/mm49/link/cache/nextflow/singularity/quay.io-sanger-tol-busco_painter-1.0.0-c2.img

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line
```

`$PATH` indeed doesn't have `/opt/conda/bin`.
How is it working in TreeVal ?

If we need a new container, can you try Seqera again ? Otherwise, if we make our own, I'd suggest using <https://hub.docker.com/_/r-base/> as the base and running the native R commands to install packages.

But also, seeing how the tool is used in multiple pipelines, I propose we do this:

1. Fork Charlotte's repository into sanger-tol (or transfer, we can ask Charlotte)
2. Add your and Will's modifications
3. Add the Merian elements TSV file there (and any other known TSV)
4. Add GitHub CI there to build a container upon release, and the container would ship the Python/R scripts

Then we have a clean busco-painter product that is independently maintained and released. And in our Nextflow pipelines we wouldn't have to keep copies of the script.
And we could even put the modules into the new [sanger-tol/nf-core-modules](https://github.com/sanger-tol/nf-core-modules)

muffato commented on workflows/genomenote.nf:
```diff
@@ -220,6 +222,22 @@ workflow GENOMENOTE {
+    //              Once there are more options, this should be reviewed for a better system
+    //
+
+    if (ch_ancestral_table) {
```
`if` doesn't work on channels
```suggestion
    if ( params.ancestral_table ) {
```

DLBPointon committed:
Merge pull request #198 from sanger-tol/dev

Dev

DLBPointon committed:
Update workflows/genomenote.nf

Co-authored-by: Matthieu Muffato <mm49@sanger.ac.uk>

DLBPointon commented:
> The R script doesn't want to run on Singularity:
> 
> ...
> 
> `$PATH` indeed doesn't have `/opt/conda/bin`. How is it working in TreeVal ?

Odd since I've had it running and the CI has succeeded with it too. We don't use the painter in treeval, once we can display PNGs in JBrowse it'll be nice. just not yet.

> 
> If we need a new container, can you try Seqera again ? Otherwise, if we make our own, I'd suggest using https://hub.docker.com/_/r-base/ as the base and running the native R commands to install packages.
> 

I did initially start with r_base with the sanger-tol container but there was something i couldn't get working so i built from scratch.

The sequera builder in the end just wouldn't work for this. Something about one of the packages not being able to be installed. ¯\_(ツ)_/¯ which is why i ended up building our own.

> But also, seeing how the tool is used in multiple pipelines, I propose we do this:
> 
> 1. Fork Charlotte's repository into sanger-tol (or transfer, we can ask Charlotte)
> 2. Add your and Will's modifications
> 3. Add the Merian elements TSV file there (and any other known TSV)
> 4. Add GitHub CI there to build a container upon release, and the container would ship the Python/R scripts
> 
> Then we have a clean busco-painter product that is independently maintained and released. And in our Nextflow pipelines we wouldn't have to keep copies of the script. And we could even put the modules into the new [sanger-tol/nf-core-modules](https://github.com/sanger-tol/nf-core-modules).

Big fan of this!

DLBPointon committed:
Merge branch 'dev' into dp24_ancestral_plots

DLBPointon committed:
Update CI temporarily

DLBPointon committed:
Prettier linting

DLBPointon commented:
Just emailed Charlotte now so we can get moving on it again.

DLBPointon commented:
I've now forked Charlotte's repo as https://github.com/sanger-tol/busco_painter

I've updated the scripts in use, to the versions in use and asked @weaglesBio to have a look at the `src/buscopainter.py` as i refactored a little. I've also added Ruff lint and formatting for the python script and have a PR open to add lintr for the R script, I'm having issues with that though as I havn't touched much R in a while and I'm not entirely sure how to implement it.

Would it be you (@muffato) or gq to add container building?

muffato commented:
Thanks for forking the repo, @DLBPointon . I've added branch protection to `main` and given write permission to the repo to Will and I.
Please ask Guoying for setting up the container build. We'll keep the container here on GitHub, similar to how he's done it for MitoHifi (and others)

DLBPointon committed:
Update main.nf

DLBPointon committed:
Update to look like sanger-tol/nf-core-module proposal

DLBPointon committed:
Update for new module structure

DLBPointon committed:
Update to turn [] into a Channel so that we can use .first(), if we don't then it will only run once or crash

DLBPointon committed:
Missing

DLBPointon committed:
Updating args

DLBPointon committed:
Update based on comments

DLBPointon committed:
Update based on comments

DLBPointon committed:
Update based on comments

DLBPointon committed:
Update to actually work

DLBPointon committed:
Corrections to args

DLBPointon committed:
UPDATE AGAIN

DLBPointon committed:
UPDATE AGAIN

DLBPointon committed:
UPDATE AGAIN

DLBPointon commented:
I've set up the container workflow, now looking at setting up the conda environment otherwise the sanger-tol/nf-core-modules workflows will crash.

DLBPointon committed:
Update to use the new container

DLBPointon committed:
Adding the nf-core/blobtk/plot module

DLBPointon committed:
Update structure for module

DLBPointon commented:
> I've set up the container workflow, now looking at setting up the conda environment otherwise the sanger-tol/nf-core-modules workflows will crash.

I'm going to skip this, I didn't realise what goes into making a conda package and the pipeline can't run in conda anyway.

Only thing that needs doing now is updating the required tests from admin.

DLBPointon committed:
Update Changelog

DLBPointon committed:
remove again

DLBPointon committed:
remove again

DLBPointon committed:
Update to better logic than '.first()'

muffato commented on CHANGELOG.md:
```diff
@@ -29,23 +32,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+| `pretextsnapshot`     |             | 0.0.4                       |
+| `samtools/faidx`      |             | 1.21                        |
+| `merquryfk`           |             | FK=38b07c2 MFK=1.1.2 R=4.42 |
+| `buscopainter.py`     |             | 1.0.0                       |
+| `plot_buscopainter.R` |             | 1.0.0                       |
```
because your using the 1.0.1 container
```suggestion
| `buscopainter.py`     |             | 1.0.1                       |
| `plot_buscopainter.R` |             | 1.0.1                       |
```

DLBPointon replied:
Done

DLBPointon committed:
Suggestion from @prototaxites

muffato commented on CHANGELOG.md:
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Addition of GenescopeFK histogram plots and related data #181 [#157](https://github.com/sanger-tol/genomenote/issues/157).
 - Support for CLI provided Busco lineage to override the NCBI auto generated one.
+- Addition of the Ancestral Element Analysis (Annotation Ancestral) subworkflow
+  - This only works for Lepidoptera (for now)
+  - The lepidoptera_annotation_table.tsv has been added
```
```suggestion
  - Refer to <https://github.com/sanger-tol/busco_painter> for more information about the tool and its configuration. That repository also contains a complete annotation table for _Lepidoptera_.
```

DLBPointon replied:
Done

muffato commented on conf/test.config:
```diff
@@ -28,6 +29,7 @@ params {
 
     // Annotation file
     annotation_set              = "https://tolit.cog.sanger.ac.uk/test-data/Ceramica_pisi/annotation/Ceramica_pisi-GCA_963859965.1-2024_04-genes.gff3.gz"
+    ancestral_table             = "https://raw.githubusercontent.com/charlottewright/Chromosome_evolution_Lepidoptera_MS/refs/tags/pub_v1/sup_tables/TableS4_Merian_element_definitions.tsv"
```
Can you put this file in https://github.com/sanger-tol/busco_painter so there are no external dependencies ?

DLBPointon replied:
Done

muffato commented on conf/test_full.config:
```diff
@@ -19,6 +19,7 @@ params {
 
     // Annotation file
     annotation_set              = "https://tolit.cog.sanger.ac.uk/test-data/Ypsolopha_sequella/annotation/GCA_934047225.1.braker.2022_09.gff3.gz"
+    ancestral_table             = "https://raw.githubusercontent.com/charlottewright/Chromosome_evolution_Lepidoptera_MS/refs/tags/pub_v1/sup_tables/TableS4_Merian_element_definitions.tsv"
```
Same here

DLBPointon replied:
Done

muffato commented:
> Only thing that needs doing now is updating the required tests from admin.

What do you mean ? The test profile works

muffato commented on bin/buscopainter.py:
Can be removed now that you have the container

DLBPointon replied:
Done

muffato commented on bin/plot_buscopainter.R:
Can be removed too

DLBPointon replied:
Done

DLBPointon committed:
Update CHANGELOG.md

Correct versioning by @muffato

Co-authored-by: Matthieu Muffato <mm49@sanger.ac.uk>

DLBPointon committed:
Merge pull request #205 from sanger-tol/blobtk_plot_update

Blobtk plot update

DLBPointon commented:
> > Only thing that needs doing now is updating the required tests from admin.
> 
> What do you mean ? The test profile works

Sorry, I mean the actual CI. Some of the jobs are for old CI.

DLBPointon committed:
Remove script that's now containerised

DLBPointon committed:
Remove script that's now containerised

DLBPointon committed:
Update https link to sanger-tol repo

DLBPointon committed:
Update comments and versions

muffato commented:
>AssertionError: We assume that software versions are the same between all modules. If you see this error-message it means you discovered an edge-case and should open an issue in nf-core/tools. 

I'm actually seeing this error locally too, when doing a fresh run (usually I do `-resume`).

I haven't checked the uniqueness yet, but there's at least one indentation issue:
```yml
"SANGERTOL_GENOMENOTE:GENOMENOTE:GENOME_METADATA:PARSE_METADATA":
    python: 3.9.5
    "SANGERTOL_GENOMENOTE:GENOMENOTE:GET_BLOBTK_PLOTS:BLOBTK_PLOT":
        blobtk: 0.7.1
```

muffato commented:
> I haven't checked the uniqueness yet, but there's at least one indentation issue:
> 
> ```yaml
> "SANGERTOL_GENOMENOTE:GENOMENOTE:GENOME_METADATA:PARSE_METADATA":
>     python: 3.9.5
>     "SANGERTOL_GENOMENOTE:GENOMENOTE:GET_BLOBTK_PLOTS:BLOBTK_PLOT":
>         blobtk: 0.7.1
> ```

That's actually the only issue. The script works if this gets indented properly.

DLBPointon commented:
> > I haven't checked the uniqueness yet, but there's at least one indentation issue:
> > ```yaml
> > "SANGERTOL_GENOMENOTE:GENOMENOTE:GENOME_METADATA:PARSE_METADATA":
> >     python: 3.9.5
> >     "SANGERTOL_GENOMENOTE:GENOMENOTE:GET_BLOBTK_PLOTS:BLOBTK_PLOT":
> >         blobtk: 0.7.1
> > ```
> 
> That's actually the only issue. The script works if this gets indented properly.

Just sorted this my end as well, i'm patching and pushing. I'll fix nf-core's version in the morning, interesting how the linting and tests didn't fail on this 

As note to self: https://github.com/nf-core/modules/pull/8998/files#diff-bfb2442cc881199c9975b625ad420c0e92a46db96c53f237fbae536c8a7d9801R48-R51

DLBPointon committed:
Tab issue

DLBPointon committed:
Patching

DLBPointon committed:
linting
